### PR TITLE
add option to overwrite ska3-* meta-package versions

### PR DESCRIPTION
**Pull request by javierggt**
This PR adds a command argument `--skare3-overwrite-version` which is meant for continuous integration to create pre-release conda packages.

It does that by:
- adding the `--skare3-overwrite-version` argument
- parsing the value provided
  - the value can be of the form `<initial-version>:<final-version>`.
  - or just a single version, in which case it assumes that the the initial version is the same os the final version but removing the release candidate part (i.e.: something that looks like "rcN" or "aN" or "bN".
- the package definition directory is copied to the temp source directory
- meta.yaml is read, all occurrences of `<initial-version>` are replaced by `<final-version>`. This is in the package version and in its dependencies, whenever their name is of the form 'ska3-*'.

# Testing

built the following:
- Quaternion (should have no effect)
- ska3-template (pkg_def dir contains a subdir)
- ska3-flight-2020.10rcN (checks the proper parsing of version as string)
- ska3-flight-2020.13rc2+shiny

### CODE REVIEW

File: ska_builder.py [hunk 1]
```diff
@@ -203,9 +213,49 @@ def build_list_packages(pkg_names, args, src_dir, build_dir):
...
+        t = jinja2.Template(fh.read())
+    text = (t.render(SKA_PKG_VERSION='$SKA_PKG_VERSION',
+                     SKA_TOP_SRC_DIR='$SKA_TOP_SRC_DIR'))
+    if version.parse(yaml.__version__) < version.parse("5.1"):
+        data = yaml.load(text)
+    else:
+        data = yaml.load(text, Loader=yaml.BaseLoader)
+    if str(data['package']['version']) == str(current_version):
+        data['package']['version'] = new_version
+        for i in range(len(data['requirements'])):
```
Comment (taldcroft) [inline ska_builder.py:228]: Changing the version in the run requirements is confusing me. 

For example, consider building `ska3-flight`. In order to get into this function we need that `current_version` matches the package version in the `ska3-flight/meta.yaml`. But then this code looks for any other run requirement package that starts with `ska3-` and if that version also matches `current_version` then it gets replaced.  So if `ska3-core` happens to have the same version as `ska3-flight` then it gets replaced?

File: ska_builder.py [hunk 2]
```diff
@@ -203,9 +213,49 @@ def build_list_packages(pkg_names, args, src_dir, build_dir):
...
+    else:
+        print(f'  - NOT changing version on {pkg_path}')
+
+
 def main():
     args = get_opt()
 
+    if args.skare3_overwrite_version:
+        if ':' not in args.skare3_overwrite_version:
+            rc = re.match('(?P<version>(?P<release>\S+)(a|b|rc)[0-9]+(\+(?P<label>\S+))?)$',
```
Comment (taldcroft) [inline ska_builder.py:249]: Use a verbose multiline form of this regex to make it easier to read.

File: ska_builder.py [hunk 3]
```diff
@@ -203,9 +213,49 @@ def build_list_packages(pkg_names, args, src_dir, build_dir):
...
+            f.write(t)
+    else:
+        print(f'  - NOT changing version on {pkg_path}')
+
+
 def main():
     args = get_opt()
 
+    if args.skare3_overwrite_version:
+        if ':' not in args.skare3_overwrite_version:
```
Comment (taldcroft) [inline ska_builder.py:248]: Add a comment about the intent of this block of code. It is entirely not obvious without reading the PR description.

File: ska_builder.py [hunk 4]
```diff
@@ -60,6 +62,8 @@ def get_opt():
                         "except on Windows")
     parser.add_argument("--repo-url",
                         help="Use this URL instead of meta['about']['home']")
+    parser.add_argument('--skare3-overwrite-version',
```
[the remark this answers is not in the document]
Comment (taldcroft) [reply ska_builder.py:65]: This argument has a complex behavior that needs documentation.

File: ska_builder.py (hunk 1 above)
Comment (javierggt) [reply ska_builder.py:228]: Yes, because the intent of this option is to replace a version by another in a skare3 release.

All ska3-* packages with matching version would have their version changed, and you would want their inter-dependencies also changed. If you are replacing the version of ska3-flight, and ska3-core appears as a dependency with the same version, then presumably you also need to build ska3-core with the replaced version.

In other words. Suppose you have ska3-core==2020.13 and ska3-flight 2020.13. Then ska3-flight will depend on ska3-core==2020.13. The way it is used is that all appearances of 2020.13 are replaced (by something like 2020.13rc1). Do you see any case where one would replace ska3-flight version and not ska3-core?

File: ska_builder.py (hunk 2 above)
Comment (javierggt) [reply ska_builder.py:249]: what do you mean by that?

File: ska_builder.py [hunk 5]
```diff
@@ -203,9 +213,49 @@ def build_list_packages(pkg_names, args, src_dir, build_dir):
         raise ValueError("Packages {} failed".format(",".join(failures)))
 
 
+def overwrite_skare3_version(current_version, new_version, pkg_path):
+    meta_file = pkg_path / 'meta.yaml'
+    with open(meta_file) as fh:
+        t = jinja2.Template(fh.read())
+    text = (t.render(SKA_PKG_VERSION='$SKA_PKG_VERSION',
+                     SKA_TOP_SRC_DIR='$SKA_TOP_SRC_DIR'))
+    if version.parse(yaml.__version__) < version.parse("5.1"):
```
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:222]: this is (was?) needed because meta.yaml files include `{{ SKA_PKG_VERSION }}`. I was trying to deal with parsing the file while leaving these unchanged. What the code does now is it replaces them by something that can be replaced back using string.Template.

The ideal would be to not replace them to begin with, but the parser was stumbling with these symbols. I will see again if there is a better option.

File: ska_builder.py [hunk 6]
```diff
@@ -111,7 +115,13 @@ def clone_repo(name, args, src_dir, meta):
 
 
 def build_package(name, args, src_dir, build_dir):
-    pkg_path = os.path.join(PKG_DEFS_PATH, name)
+    pkg_path = Path(src_dir) / 'pkg_defs' / name
```
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:118]: src_dir is actually a temporary directory which gets created right before calling the function. Why wouldn't it be empty when it starts?

It used to be jusr `./src`, relative to the working directory, but now it is a temp dir per process, so it starts empty and is populated with subdirs called by each package's name. Where is the collision?

File: ska_builder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:228]: this is interesting... one of those "how come this has worked before?". It should be all requirements, build, run, whatever. I'm checking why this ever worked.

File: ska_builder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:222]: This was a case of things working by coincidence, because we always put the ska3-* meta-packages at the top of the requirement list. Committing fix now.

File: ska_builder.py [hunk 7]
```diff
@@ -111,7 +115,13 @@ def clone_repo(name, args, src_dir, meta):
...
+    shutil.copytree(PKG_DEFS_PATH / name, pkg_path)
+
+    if args.skare3_overwrite_version and re.match('ska3-\S+$', name):
```
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:121]: Fixed.

File: ska_builder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:65]: Done.

File: ska_builder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:222]: I removed this. If we need to, it can always be added.

File: ska_builder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:228]: This is a case where things were working by coincidence, only because we always put the ska3-* meta-packages at the top of the requirement list.

File: ska_builder.py [hunk 8]
```diff
@@ -203,9 +213,49 @@ def build_list_packages(pkg_names, args, src_dir, build_dir):
...
+    text = (t.render(SKA_PKG_VERSION='$SKA_PKG_VERSION',
+                     SKA_TOP_SRC_DIR='$SKA_TOP_SRC_DIR'))
+    if version.parse(yaml.__version__) < version.parse("5.1"):
+        data = yaml.load(text)
+    else:
+        data = yaml.load(text, Loader=yaml.BaseLoader)
+    if str(data['package']['version']) == str(current_version):
+        data['package']['version'] = new_version
+        for i in range(len(data['requirements'])):
+            if re.search('==', data['requirements']['run'][i]):
```
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py:229]: Done

File: ska_builder.py (hunk 3 above)
Comment (javierggt) [reply ska_builder.py:248]: Done.

File: ska_builder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (taldcroft) [reply ska_builder.py]: I meant the yaml version check.

File: ska_builder.py (hunk 1 above)
Comment (taldcroft) [reply ska_builder.py]: But does this mean that version numbers of ska3-core and ska3-flight need to track each other in order for this scheme to work? I may still just be generally confused about the whole workflow.

File: ska_builder.py (hunk 2 above)
Comment (taldcroft) [reply ska_builder.py]: Like the example here: https://docs.python.org/3/library/re.html#re.VERBOSE

File: ska_builder.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (taldcroft) [reply ska_builder.py:123]: Ah OK I was misthinking that `src_dir` was the root of the package git repo (which might include `pkg_defs` by chance). But never mind.

File: ska_builder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py]: ah, I guess it is not needed anymore because we are using the newer version of yaml anyway.

File: ska_builder.py (hunk 1 above)
Comment (javierggt) [reply ska_builder.py]: No. You can release ska3-flight 2020.13rc1 which depends on ska3-core 2020.12. This option would replace appearances of 2020.13 by 2020.13rc1, and ska3-core does not match, so it is unchanged.

If you were to update both of them, both ska3-flight/core are 2020.13rc1, then ska3-flight must depend on ska3-core 2020.13, so the ska3-core version in the requirements is also replaced by 2020.13rc1.

File: ska_builder.py (hunk 2 above)
Comment (javierggt) [reply ska_builder.py]: ah, ok. I just made some changes for this.

File: ska_builder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (javierggt) [reply ska_builder.py]: I removed the check.

**Comment by javierggt** [comment]
@taldcroft and @jeanconn, I took some time to think about this and about an issue I mentioned in slack that relates to this PR. The situation is as follows.

Conda uses *[preprocessing selectors](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#preprocessing-selectors)* within square brackets to define metadata. A selector has the form `# [<selector>]` at the end of a line. These selectors are written within comments. So far so good. Nothing new.

Now... the process to overwrite the version loads the `meta.yaml` file, replaces the version, and then writes it. In the process, all comments (together with all preprocessing selectors) are discarded. That is a problem.

No solution is perfect. The options are:
- do not load/write yaml, but instead read the file line by line, do the replacements, and write (no yaml). If one does this, there are two options:
  - adopt conventions in our meta.yaml files to make them easier to parse and search for versions. For example, to never write
     ```
     version:
       2020.13
     ```
     but instead write
     ```
     version: 2020.13
     ```
  - basically handle the yaml format by hand. **This out of the question.**
- Apply the same filtering conda applies, but before calling conda. That could be possible using [conda_build.metadata.Metadata](https://github.com/conda/conda-build/blob/master/conda_build/metadata.py).
- Scratch the whole idea of overwriting the version for pre-releases and rethinking the release+test whole workflow.

I don't know about you, but I prefer the first option. The last change I added does that.

**Comment by jeanconn** [comment]
I suppose I'm still a little confused.  We've gotten by up to this point because ska3-flight is not arch-specific, right?  So the issue is with ska3-core and ska3-perl.  The plan is to update ska3-core on like a yearly basis and basically never update ska3-perl, so maybe we are OK?

**Comment by javierggt** [comment]
yes, but if you automate things in a workflow you would automate them for all. Do you think when we update ska3-core we will not do a release candidate? So we would have to treat them separately?

For example, today I tagged the release candidate rc3 of version 2020.13+shiny and that caused all packages to be built. That is ska3-core/flight for windows/linux/OSX. These builds are wrong, of course, because of the issue with selectors, as described above, but that is exactly the reason for having release candidates.

**Comment by javierggt** [comment]
So, it all boils down to a simple question:

> Are you willing to take the convention of always writing our meta.yaml for ska3-\* packages so that the version specification is not split in two lines?

**Comment by jeanconn** [comment]
I thought perhaps we could skip release candidates on ska3-core the way that we currently skip them on every package except ska3-flight/ska3-matlab... the promotion process is controlled by the ska3-flight package, which *would* have a release candidate versio.

**Comment by javierggt** [comment]
I do not know what you mean. ska3-core packages are tagged with the corresponding ska3-* version. That it does not go through load review is another thing. And whenever you tag, there will be an iteration with testing until you get it right and then you make the final tag.

**Comment by jeanconn** [comment]
"ska3-core packages are tagged with the corresponding ska3-* version." I thought that was still work-in-progress and we had not yet agreed on that convention.  I love the RC candidates and workflow, but I'm not convinced that ska3-flight should match ska3-core versions.  Also will probably need "Are you willing to take the convention of always writing our meta.yaml for ska3-* packages so that the version specification is not split in two lines?" explained to me again.

**Comment by javierggt** [comment]
> Also will probably need "Are you willing to take the convention of always writing our meta.yaml for ska3-* packages so that the version specification is not split in two lines?" explained to me again.

never write:
```
     version:
       2020.13
```
but instead write
```
     version: 2020.13
```

**Comment by jeanconn** [comment]
If that is the only gotcha of this system, I'm fine with always writing a one-line version specification.
